### PR TITLE
SPEC-173: Document tests conditional on server version

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/SPEC-173

Note: I'll commit the generated JSON files once the YAML tests are reviewed. I ran these tests against the PHP library (without my [PHPLIB-95](https://jira.mongodb.org/browse/PHPLIB-95) work-around in place) on server versions 2.4, 2.6, and 3.4 without issue.

### CODE REVIEW

File: source/crud/tests/write/findOneAndReplace-upsert.yml [hunk 1]
```diff
@@ -3,18 +3,20 @@ data:
...
             name: findOneAndReplace
             arguments:
                 filter: {_id: 4}
                 replacement: {x: 44}
                 projection: {x: 1, _id: 0}
-                sort: {x: 1}
+                # Omit the sort option as it has no effect when no documents
+                # match and would only cause an inconsistent return value on
+                # pre-3.0 servers when combined with returnDocument "before"
+                # (see: SERVER-17650).
```
Comment (jmikola) [inline source/crud/tests/write/findOneAndReplace-upsert.yml]: Per the comment, the `sort` option had no purpose here other than creating an inconsistent result value for pre-3.0 servers. It seemed prudent to simply remove it and make the test more compatible, rather than create test variants.

File: source/crud/tests/write/findOneAndReplace-upsert.yml [hunk 2]
```diff
@@ -45,3 +47,45 @@ tests:
                     - {_id: 2, x: 22}
                     - {_id: 3, x: 33}
                     - {_id: 4, x: 44}
+    -
+        description: "FindOneAndReplace when no documents match with id specified with upsert returning the document before modification"
```
Comment (jmikola) [inline source/crud/tests/write/findOneAndReplace-upsert.yml]: The two tests "without id specified" above previously existed in `findOneAndReplace.yml` but these "with id specified" variants are new. This makes the test cases consistent with what we already had in `replaceOne.yml` for the `update` command.

File: source/crud/tests/write/findOneAndReplace-upsert_pre_2.6.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
+data:
+    - {_id: 1, x: 11}
+    - {_id: 2, x: 22}
+    - {_id: 3, x: 33}
+# This file includes the same test cases as findOneAndReplace-upsert.yml with
+# some omissions for pre-2.6 servers. We cannot verify the ID of an upserted
+# document in some cases due to SERVER-5289.
+maxServerVersion: '2.4.99'
```
Comment (jmikola) [inline source/crud/tests/write/findOneAndReplace-upsert_pre_2.6.yml]: Using "X.Y" for `minServerVersion` isn't really an issue, as we can assume the minor version is zero. That doesn't work as nicely for `maxServerVersion`, so we can either use the next minor version or a sufficiently high patch version in the same "X.Y" series. The [CRUD test README](https://github.com/mongodb/specifications/blob/master/source/crud/tests/README.rst) does not specify whether `minServerVersion` and `maxServerVersion` are inclusive or exclusive, so I opted to use a high patch version instead of "2.5" in this case.

On the inclusive/exclusive front, I suppose that's only an edge case when comparing a test's `minServerVersion` against a server with a "0" patch version (e.g. "2.6.0"). To that end, we should probably document that these ranges are inclusive. I opened [SPEC-834](https://jira.mongodb.org/browse/SPEC-834) to track this.

File: source/crud/tests/write/findOneAndReplace.yml [hunk 4]
```diff
@@ -94,20 +94,6 @@ tests:
                     - {_id: 2, x: 22}
                     - {_id: 3, x: 33}
     -
-        description: "FindOneAndReplace when no documents match with upsert returning the document before modification"
```
Comment (jmikola) [inline source/crud/tests/write/findOneAndReplace.yml]: These two tests were moved to `findOneAndReplace-upsert.yml`.

File: source/crud/tests/write/replaceOne-collation.yml [hunk 5]
```diff
@@ -17,6 +17,7 @@ tests:
             result:
                 matchedCount: 1
                 modifiedCount: 1
+                upsertedCount: 0
```
Comment (jmikola) [inline source/crud/tests/write/replaceOne-collation.yml]: Previously, no tests were asserting `upsertedCount`. It seems like a freebie, since the field is available for all server versions.

File: source/crud/tests/write/replaceOne-pre_2.6.yml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                replacement: {x: 1}
+                upsert: true
+
+        outcome:
+            result:
+                matchedCount: 0
+                upsertedCount: 1
+            # Can't verify upsertedId or collection data because server versions
+            # before 2.6 do not take the _id from the filter document during an
+            # upsert (see: SERVER-5289)
```
Comment (jmikola) [inline source/crud/tests/write/replaceOne-pre_2.6.yml]: I'll note that without the addition of `upsertedCount` here, this test wouldn't really be asserting anything meaningful due to [SERVER-5289](https://jira.mongodb.org/browse/SERVER-5289).

File: source/crud/tests/write/replaceOne-upsert.yml [hunk 7]
```diff
@@ -1,48 +0,0 @@
-data:
-    - {_id: 1, x: 11}
-    - {_id: 2, x: 22}
-    - {_id: 3, x: 33}
-minServerVersion: '2.6'
-# See SERVER-5289 for why the collection data is only checked for server versions >= 2.6
```
Comment (jmikola) [inline source/crud/tests/write/replaceOne-upsert.yml]: Splitting off `replaceOne-pre_2.6.yml` allowed us to relocate the upsert tests to the base `replaceOne.yml` file and remove this entirely.

**Comment by xdg** [comment]
I only have time this week for a quick skim, but what I saw seemed sensible.

**Comment by craiggwilson** [comment]
Looks great Jeremy.  lgtm.
